src/App.tsx

```diff
@@ -1,24 +1,190 @@
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isDisable, setIsDisable] = useState<boolean>(false);
+  const [loadingItems, setLoadingItems] = useState<number[]>([]);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  const setErrorMesssage = (message: string) => {
+    if (message) {
+      setErrorMesssageState(message);
+    } else {
+      setErrorMesssageState('');
+    }
+  };
+
+  const handleTodoClick = (todoId: number) => {
+    const clickedTodo = todos.find((todo) => todo.id === todoId);
+
+    if (clickedTodo) {
+      updateTodos({
+        id: todoId,
+        completed: !clickedTodo.completed,
+      })
+        .then((response) => {
+          setTodos(prevTodos => {
+            return prevTodos.map((todo) => {
+              if (todo.id === response.id) {
+                return response as Todo;
+              }
+
+              return todo;
+            });
+          });
+        })
+        .catch(() => setErrorMesssage(Errors.update));
+    }
+  };
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos(USER_ID)
+      .then((fetchedTodos) => {
+        setTodos(fetchedTodos);
+      })
+      .catch(() => setErrorMesssage(Errors.load))
+      .finally(() => {
+        setIsLoading(false);
+      });
+  }, []);
+
+  const filteredTodos = useMemo(() => [...todos].filter(({ completed }) => {
+    if (filter === Filters.All) {
+      return true;
+    }
+
+    return (filter === Filters.Active)
+      ? !completed
+      : completed;
+  }), [todos, filter]);
+
+  const removeTodo = (todoId: number) => {
+    setIsLoading(true);
+    setLoadingItems(current => [...current, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => setTodos(current => current
+        .filter(todo => todo.id !== todoId)))
+      .catch(() => setErrorMesssage(Errors.delete))
+      .finally(() => {
+        setIsLoading(false);
+        // eslint-disable-next-line max-len
+        setLoadingItems(current => current.filter((id: number) => id !== todoId));
+      });
+  };
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!title.trim()) {
+      return setErrorMesssage(Errors.title);
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: title.trim(),
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+
+    setIsDisable(true);
+
+    return addTodo(newTodo)
+      .then((resultTodo:Todo) => {
+        setTodos(current => [...current, resultTodo]);
+        setTitle('');
+      })
+      .catch(() => setErrorMesssage(Errors.add))
+      .finally(() => {
+        setIsDisable(false);
+        setTempTodo(null);
+      });
+  };
+
+  const clearCompletedTodos = () => {
+    // eslint-disable-next-line max-len
+    const completedTodoIds = todos.filter(({ completed }) => completed).map(({ id }) => id);
+
+    completedTodoIds.forEach((id) => removeTodo(id));
+  };
+
+  const handleToggleAll = () => {
```

Comment:
One of the tests is failing because we don't set loading state to true here. 
We should show loader immediately after pressing a toggle button, before we actually get the response from the server.